File: JavaLibrary/src/test/java/com/flat/simplex/parser/logic/line/call/CallStructTest.java
package com.flat.simplex.parser.logic.line.call;

import com.flat.simplex.parser.logic.Context;
import com.flat.simplex.parser.logic.block.BlockIf;
import com.flat.simplex.parser.logic.error.Error;
import com.flat.simplex.support.TokenChain;
import org.junit.jupiter.api.Test;

import static com.flat.simplex.support.ContextSupport.assertErrors;
import static com.flat.simplex.support.TokenChain.readChain;
import static org.junit.jupiter.api.Assertions.*;

class CallStructTest {

    @Test
    public void loadEmpty() {
        TokenChain chain = readChain("{}");
        Context context = new Context(chain.get());
        CallStruct call = new CallStruct(context, chain.get());
        call.load();

        assertEquals(0, call.getMembers().size(), "Invalid members size");

        assertErrors(context);
    }

    @Test
    public void loadSingle() {
        TokenChain chain = readChain("{ a : b }");
        Context context = new Context(chain.get());
        CallStruct call = new CallStruct(context, chain.get());
        call.load();

        assertEquals(1, call.getMembers().size(), "Invalid members size");

        assertErrors(context);
    }

    @Test
    public void loadMultiple() {
        TokenChain chain = readChain("{ a : b, c : d, e : f }");
        Context context = new Context(chain.get());
        CallStruct call = new CallStruct(context, chain.get());
        call.load();

        assertEquals(3, call.getMembers().size(), "Invalid members size");

        assertErrors(context);
    }

    @Test
    public void loadSingleOperation() {
        TokenChain chain = readChain("{ a : b + c }");
        Context context = new Context(chain.get());
        CallStruct call = new CallStruct(context, chain.get());
        call.load();

        assertEquals(1, call.getMembers().size(), "Invalid members size");

        assertErrors(context);
    }

    @Test
    public void loadSingleTernary() {
        TokenChain chain = readChain("{ a : b ? c : d }");
        Context context = new Context(chain.get());
        CallStruct call = new CallStruct(context, chain.get());
        call.load();

        assertEquals(1, call.getMembers().size(), "Invalid members size");

        assertErrors(context);
    }

    @Test
    public void loadString_Fail() {
        TokenChain chain = readChain("{ \"a\" : b }");
        Context context = new Context(chain.get());
        CallStruct call = new CallStruct(context, chain.get());
        call.load();

        assertEquals(1, call.getMembers().size(), "Invalid members size");

        assertErrors(context, Error.structDoNotUseString);
    }

    @Test
    public void loadMissingColon_Fail() {
        TokenChain chain = readChain("{a b}");
        Context context = new Context(chain.get());
        CallStruct call = new CallStruct(context, chain.get());
        call.load();

        assertEquals(0, call.getMembers().size(), "Invalid members size");

        assertErrors(context, Error.unexpectedToken, Error.unexpectedEndOfTokens);
    }

    @Test
    public void loadMissingEnd_Fail() {
        TokenChain chain = readChain("{a : }");
        Context context = new Context(chain.get());
        CallStruct call = new CallStruct(context, chain.get());
        call.load();

        assertEquals(0, call.getMembers().size(), "Invalid members size");

        assertErrors(context, Error.unexpectedEndOfTokens);
    }

    @Test
    public void loadMissingAfterComma_Fail() {
        TokenChain chain = readChain("{a : b, }");
        Context context = new Context(chain.get());
        CallStruct call = new CallStruct(context, chain.get());
        call.load();

        assertEquals(1, call.getMembers().size(), "Invalid members size");

        assertErrors(context, Error.unexpectedEndOfTokens);
    }

    @Test
    public void loadMissingCloser_Fail() {
        TokenChain chain = readChain("{ a : b");
        Context context = new Context(chain.get());
        CallStruct call = new CallStruct(context, chain.get());
        call.load();

        assertEquals(1, call.getMembers().size(), "Invalid members size");

        assertErrors(context, Error.missingCloser);
    }
}

File: SimplexApp/simplex/parser/logic/line/call/CallStruct.test.js
const TokenChain = require("simplex/support/TokenChain.js");
const Context = require("simplex/parser/logic/Context.js");
const ContextSupport = require("simplex/support/ContextSupport.js");
const Error = require("simplex/parser/logic/error/Error.js");
const CallStruct = require("simplex/parser/logic/line/call/CallStruct.js");
const AssertError = require("simplex/support/AssertError.js");

const readChain = TokenChain.readChain;
const assertErrors = ContextSupport.assertErrors;
const assertEquals = AssertError.assertEquals;

test('loadEmpty', () => {
    let chain = readChain('{}');
    let context = new Context(chain.get());
    let call = new CallStruct(context, chain.get());
    call.load();
    assertEquals(0, call.getMembers().length, 'Invalid members size');
    assertErrors(context);
});
test('loadSingle', () => {
    let chain = readChain('{ a : b }');
    let context = new Context(chain.get());
    let call = new CallStruct(context, chain.get());
    call.load();
    assertEquals(1, call.getMembers().length, 'Invalid members size');
    assertErrors(context);
});
test('loadMultiple', () => {
    let chain = readChain('{ a : b, c : d, e : f }');
    let context = new Context(chain.get());
    let call = new CallStruct(context, chain.get());
    call.load();
    assertEquals(3, call.getMembers().length, 'Invalid members size');
    assertErrors(context);
});
test('loadSingleOperation', () => {
    let chain = readChain('{ a : b + c }');
    let context = new Context(chain.get());
    let call = new CallStruct(context, chain.get());
    call.load();
    assertEquals(1, call.getMembers().length, 'Invalid members size');
    assertErrors(context);
});
test('loadSingleTernary', () => {
    let chain = readChain('{ a : b ? c : d }');
    let context = new Context(chain.get());
    let call = new CallStruct(context, chain.get());
    call.load();
    assertEquals(1, call.getMembers().length, 'Invalid members size');
    assertErrors(context);
});
test('loadString_Fail', () => {
    let chain = readChain('{ \'a\' : b }');
    let context = new Context(chain.get());
    let call = new CallStruct(context, chain.get());
    call.load();
    assertEquals(1, call.getMembers().length, 'Invalid members size');
    assertErrors(context, Error.structDoNotUseString);
});
test('loadMissingColon_Fail', () => {
    let chain = readChain('{a b}');
    let context = new Context(chain.get());
    let call = new CallStruct(context, chain.get());
    call.load();
    assertEquals(0, call.getMembers().length, 'Invalid members size');
    assertErrors(context, Error.unexpectedToken, Error.unexpectedEndOfTokens);
});
test('loadMissingEnd_Fail', () => {
    let chain = readChain('{a : }');
    let context = new Context(chain.get());
    let call = new CallStruct(context, chain.get());
    call.load();
    assertEquals(0, call.getMembers().length, 'Invalid members size');
    assertErrors(context, Error.unexpectedEndOfTokens);
});
test('loadMissingAfterComma_Fail', () => {
    let chain = readChain('{a : b, }');
    let context = new Context(chain.get());
    let call = new CallStruct(context, chain.get());
    call.load();
    assertEquals(1, call.getMembers().length, 'Invalid members size');
    assertErrors(context, Error.unexpectedEndOfTokens);
});
test('loadMissingCloser_Fail', () => {
    let chain = readChain('{ a : b');
    let context = new Context(chain.get());
    let call = new CallStruct(context, chain.get());
    call.load();
    assertEquals(1, call.getMembers().length, 'Invalid members size');
    assertErrors(context, Error.missingCloser);
});